model/state_exec_timeout.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2022 The Serverless Workflow Specification Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	validator "github.com/go-playground/validator/v10"
+	val "github.com/serverlessworkflow/sdk-go/v2/validator"
+)
+
+// StateExecTimeout defines workflow state execution timeout
+type StateExecTimeout struct {
+	// Single state execution timeout, not including retries (ISO 8601 duration format)
+	Single string `json:"single,omitempty"`
+	// Total state execution timeout, including retries (ISO 8601 duration format)
+	Total string `json:"total" validate:"required"`
+}
+
+// nolint
+// just define another type to unmarshal object, so the UnmarshalJSON will not called recursively
+type stateExecTimeoutForUnmarshal StateExecTimeout
+
+// UnmarshalJSON unmarshal StateExecTimeout object from json bytes
+func (s *StateExecTimeout) UnmarshalJSON(data []byte) error {
+	// We must trim the leading space, because we use first byte to detect data's type
+	data = bytes.TrimSpace(data)
+	if len(data) == 0 {
+		// TODO: Normalize error messages
+		return fmt.Errorf("no bytes to unmarshal")
+	}
+
+	var err error
+	switch data[0] {
+	case '"':
+		s.Total, err = unmarshalString(data)
+		return err
+	case '{':
+		var v stateExecTimeoutForUnmarshal
```

Comment:
Since this type is essentially `StateExecTimeout`, the `json.Unmarshal` below will still call this very same function, no? Not sure why we need this other type. `var v StateExecTimeout` would do.

This is ok, we call fewer unmarshals; it's a quick win.

Reply:
`stateExecTimeoutForUnmarshal` is another type which didn't have UnmarshalJSON method（didn't implement json.Unmarshaler interface）, so the UnmarshalJSON function will not recursively called.

Reply:
Please see https://go.dev/ref/spec#Type_declarations Type definitions section for more detail.

```
A defined type may have [methods](https://go.dev/ref/spec#Method_declarations) associated with it. It does not inherit any methods bound to the given type
```